panel/browser-panel-client.cpp

```diff
@@ -186,6 +194,73 @@ bool QCefBrowserClient::OnBeforePopup(
 	return true;
 }
 
+void QCefBrowserClient::OnBeforeContextMenu(CefRefPtr<CefBrowser>,
+					    CefRefPtr<CefFrame>,
+					    CefRefPtr<CefContextMenuParams>,
+					    CefRefPtr<CefMenuModel> model)
+{
+	if (model->IsVisible(MENU_ID_BACK) &&
+	    model->IsVisible(MENU_ID_FORWARD) &&
+	    (!model->IsVisible(MENU_ID_RELOAD) &&
+	     !model->IsVisible(MENU_ID_RELOAD_NOCACHE))) {
+		model->InsertItemAt(2, MENU_ID_RELOAD_NOCACHE, "&Reload");
+		if (model->IsVisible(MENU_ID_PRINT)) {
+			model->Remove(MENU_ID_PRINT);
+		}
+	}
+}
+
+bool QCefBrowserClient::RunContextMenu(
+	CefRefPtr<CefBrowser>, CefRefPtr<CefFrame>,
+	CefRefPtr<CefContextMenuParams>, CefRefPtr<CefMenuModel> model,
+	CefRefPtr<CefRunContextMenuCallback> callback)
+{
+	std::vector<std::tuple<std::string, int, bool, int>> menu_items;
+	menu_items.reserve(model->GetCount());
+	for (int i = 0; i < model->GetCount(); i++) {
+		menu_items.push_back(
+			{model->GetLabelAt(i), model->GetCommandIdAt(i),
+			 model->IsEnabledAt(i), model->GetTypeAt(i)});
+	}
+
+	QMetaObject::invokeMethod(
+		QCoreApplication::instance()->thread(),
+		[menu_items, callback]() {
+			QMenu contextMenu;
+			std::string name;
+			int command_id;
+			bool enabled;
+			int type_id;
+
+			for (int i = 0; i < menu_items.size(); i++) {
+				std::tie(name, command_id, enabled, type_id) =
+					menu_items[i];
+				switch (type_id) {
+				case MENUITEMTYPE_COMMAND: {
+					QAction *item =
+						new QAction(name.c_str());
+					item->setEnabled(enabled);
+					item->setProperty("cmd_id", command_id);
+					contextMenu.addAction(item);
+				} break;
+				case MENUITEMTYPE_SEPARATOR:
+					contextMenu.addSeparator();
+					break;
+				}
+			}
+
+			QAction *action = contextMenu.exec(QCursor::pos());
+			if (action) {
+				QVariant cmdId = action->property("cmd_id");
+				callback.get()->Continue(cmdId.toInt(),
+							 EVENTFLAG_NONE);
+			} else {
+				callback.get()->Cancel();
+			}
+		});
```

Comment:
Discussed with @WizardCM , but I'd split this into a separate function and call it similar to how we use `QMetaObject::invokeMethod` elsewhere.  That should help with indenting and line length.

Reply:
I attempted this, but `invokeMethod` from a non-Qt thread can only invoke slots/signals from QThread.

Reply:
Is this then creating an anonymous slot/signal?

Reply:
Pretty much?